### PR TITLE
[Images] Add a lock-file for mlrun/mlrun container image

### PR DESCRIPTION
**Pull request by liranbg**
Continuing the work started by https://github.com/mlrun/mlrun/pull/6677 - adding mlrun/mlrun image to the party.

This PR also brings
1. removes uv from the final image
2. Utilizing the docker host cache to optimize builds
3. Uses `--from` directive instead of ADD which has better caching optimization by docker

Next in line are mlrun/base, mlrun/test* images, mlrun/jupyter, mlrun/mlrun-gpu , mlrun/common

https://iguazio.atlassian.net/browse/ML-8382

### CODE REVIEW

File: dockerfiles/mlrun-api/Dockerfile [hunk 1]
```diff
@@ -41,21 +44,29 @@ RUN apt-get update && apt-get install --no-install-recommends -y \
...
 
-RUN sh /uv-installer.sh && rm /uv-installer.sh
+WORKDIR /mlrun
 
-ENV PATH="/root/.local/bin/:$PATH"
-ENV UV_SYSTEM_PYTHON=true
+RUN --mount=from=uv-image,source=/uv,target=/bin/uv \
+    --mount=type=cache,target=/root/.cache/uv \
+    --mount=type=bind,source=dockerfiles/mlrun-api/locked-requirements.txt,target=locked-requirements.txt \
+    uv pip sync --require-hashes locked-requirements.txt
```
Comment (TomerShor) [inline dockerfiles/mlrun-api/Dockerfile:54]: if this ( + the UV image initialization) is going to be the same in all mlrun images, perhaps we can create a custom base image that will be used as a `FROM` for all of them?
not critical for this PR.

Comment (liranbg) [reply dockerfiles/mlrun-api/Dockerfile:54]: that needs to have a bigger refactor but indeed could be optimized using bytecode compilation

File: Makefile [hunk 2]
```diff
@@ -804,8 +807,22 @@ verify-uv-version:
...
+upgrade-mlrun-mlrun-deps-lock: verify-uv-version ## Upgrade mlrun-mlrun locked requirements file
+	uv pip compile \
+		requirements.txt \
+		extras-requirements.txt \
+		dockerfiles/mlrun/requirements.txt \
+		$(MLRUN_UV_UPGRADE_FLAG) \
+		--output-file dockerfiles/mlrun/locked-requirements.txt
+
+.PHONY: upgrade-mlrun-mlrun-deps-lock
+upgrade-mlrun-deps-lock: verify-uv-version ## Upgrade mlrun-* locked requirements file
```
Comment (jond01) [inline Makefile:826]: The phony name doesn't match the target

File: dockerfiles/mlrun-api/Dockerfile [hunk 3]
```diff
@@ -13,6 +13,9 @@
 # limitations under the License.
 
 ARG MLRUN_PYTHON_VERSION=3.9
+ARG MLRUN_UV_IMAGE=ghcr.io/astral-sh/uv:latest
```
Comment (jond01) [inline dockerfiles/mlrun-api/Dockerfile:16]: Don't we want to pin it also in the Dockerfile? It's more robust than taking the latest version.

File: dockerfiles/mlrun/Dockerfile [hunk 4]
```diff
@@ -21,14 +24,6 @@ RUN apt-get update && \
     DEBIAN_FRONTEND=noninteractive apt-get -y upgrade && \
     rm -rf /var/lib/apt/lists/*
 
-# Update base pip packages (usually pip, setuptools, wheel):
-RUN python -m pip  list --disable-pip-version-check --format freeze \
```
Comment (jond01) [inline dockerfiles/mlrun/Dockerfile:25]: Don't we still want pip to be upgraded?

Comment (liranbg) [reply dockerfiles/mlrun/Dockerfile:25]: happen on line 64

**Pull request merged**